small improvements to reduce time spent finding splits

fabuzaid21 opened the pull request:
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)


## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)


(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala:
```diff
@@ -56,6 +58,34 @@ private[ml] object AltDT extends Logging {
       val minInfoGain: Double,
       val impurity: Impurity) extends Serializable {
 
+    private val unorderedSplits = mutable.Map.empty[Int, Array[CategoricalSplit]]
```
This should go in the class, not the object.  Also, is there really any benefit to computing this lazily?  It would be simpler to compute it once during the initialization.

jkbradley replied:
(Before, I missed that this was in the metadata class; sorry!)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala:
```diff
@@ -434,10 +466,10 @@ private[ml] object AltDT extends Logging {
       _ => metadata.createImpurityAggregator())
     var i = 0
-    val len = values.length
+    val len = to - from
```
Here, it would be better to iterate from i = from to i = to so that you avoid extra additions on each iteration.  Same in a few places below.

jkbradley commented:
This looks good.  I just had small comments.

fabuzaid21 commented:
Okay, just pushed some new changes addressing the feedback. I believe the unorderedSplits stuff was already in the AltDTMetadata class – did you want me to move it somewhere else?

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala:
```diff
@@ -54,30 +57,67 @@ private[ml] object AltDT extends Logging {
+      val categoricalFeaturesInfo: Map[Int, Int]) extends Serializable {
 
-    def isClassification: Boolean = numClasses >= 2
+    private val unorderedFeatures = new mutable.HashMap[Int, Array[CategoricalSplit]]()
```
We don't need this to be mutable if we do not construct it lazily.  I'd make it a non-mutable map.

Also, I'd keep the named unorderedSplits since it is more accurate.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala:
```diff
@@ -54,30 +57,67 @@ private[ml] object AltDT extends Logging {
 
-    def isMulticlass: Boolean = numClasses > 2
+    /**
+     * borrowed from [[DecisionTreeMetadata.buildMetadata]], lines 142-162
```
No need to list line numbers since those will get out of date

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/AltDT.scala:
```diff
@@ -54,30 +57,67 @@ private[ml] object AltDT extends Logging {
+      splits
+    }
+
+    def getUnorderedSplits(featureIndex: Int): Array[CategoricalSplit] = {
```
Do you need this method any more?  The require() should not be necessary unless there is a bug in the code which uses this method, and the getOrElse should not be needed since this is initialized already.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/AltDTSuite.scala:
```diff
@@ -213,35 +213,37 @@ class AltDTSuite extends SparkFunSuite with MLlibTestSparkContext  {
     val toOffset = 4
     val impurity = Entropy
-    val metadata = new AltDTMetadata(numClasses = 2, maxBins = 4, minInfoGain = 0.0, impurity)
+    val metadata = new AltDTMetadata(numClasses = 2, maxBins = 4, minInfoGain = 0.0, impurity, Map(1 -> 3))
```
Which of these changes to the test suite were necessary to fix the tests?  I would not have expected changes to be necessary, except for needing to specify categoricalFeaturesInfo

fabuzaid21 replied:
I also had to change the values to be Arrays instead of Seqs, since chooseSplit no longer to Seqs as arguments

jkbradley commented:
> I believe the unorderedSplits stuff was already in the AltDTMetadata class – did you want me to move it somewhere else?

No, that was me not expanding the Github code view.  I thought it was in the AltDT object and did not see it was in the AltDTMetadata class.

jkbradley commented:
LGTM, merging
Thanks!
